spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala

```diff
@@ -0,0 +1,181 @@
+  }
+
+  override def insert(data: DataFrame, overwrite: Boolean): Unit = {
+    if (overwrite) {
+      connector.withSessionDo {
+        val keyspace = quoted(tableRef.keyspace)
+        val table = quoted(tableRef.table)
+        session => session.execute(s"TRUNCATE $keyspace.$table")
+      }
+    }
+
+    implicit val rwf = SqlRowWriter.Factory
+    data.rdd.saveToCassandra(tableRef.keyspace, tableRef.table, AllColumns, writeConf)
+  }
+
+  override def sizeInBytes: Long = tableSizeInBytes
+
+  implicit val cassandraConnector = connector
+  implicit val readconf = readConf
+  private[this] val baseRdd =
+    sqlContext.sparkContext.cassandraTable[CassandraSQLRow](tableRef.keyspace, tableRef.table)
+
+  def buildScan() : RDD[Row] = baseRdd.asInstanceOf[RDD[Row]]
+
+  override def buildScan(requiredColumns: Array[String], filters: Array[Filter]): RDD[Row] = {
+    logInfo(s"filters: ${filters.mkString(", ")}")
+    val pushdownFilters = if(filterPushdown) new PushDown(filters, tableDef).toPushDown else Seq.empty
+    if(filterPushdown) {
+      logInfo(s"pushdown filters: ${pushdownFilters.toString()}")
+    }
+
+    val transformer = new RddTransformer(requiredColumns, pushdownFilters)
+    transformer.transform(baseRdd).asInstanceOf[RDD[Row]]
+  }
+
+  /** Quote name */
+  private[this] def quoted(str: String): String = {
+    "\"" + str + "\""
+  }
+
+  /** Add column selection and where clauses on top of baseRdd */
+  private[this] class RddTransformer(requiredColumns: Array[String], filters: Seq[Filter]) {
+
+    /** CassandraRDD[CassandraSQLRow] is the only type supported for transferring */
+    private type RDDType = CassandraRDD[CassandraSQLRow]
+
+    /** Transfer selection to limit to columns specified */
+    def maybeSelect(rdd: RDDType) : RDDType = {
+      if (requiredColumns.nonEmpty) {
+        rdd.select(requiredColumns.map(column => column: NamedColumnRef): _*)
+      } else {
+        rdd
+      }
+    }
+
+    /** Push down filters to CQL query */
+    private def maybePushdownFilters(rdd: RDDType) : RDDType = {
+      whereClause(filters) match {
+        case (cql, values) if values.nonEmpty => rdd.where(cql, values: _*)
+        case _ => rdd
+      }
+    }
+
+    /** Construct Cql clause and retrieve the values from filter */
+    private def filterToCqlAndValue(filter: Any): (String, Seq[Any]) = {
+      filter match {
+        case sources.EqualTo(attribute, value)            => (s"${quoted(attribute)} = ?", Seq(value))
+        case sources.LessThan(attribute, value)           => (s"${quoted(attribute)} < ?", Seq(value))
+        case sources.LessThanOrEqual(attribute, value)    => (s"${quoted(attribute)} <= ?", Seq(value))
+        case sources.GreaterThan(attribute, value)        => (s"${quoted(attribute)} > ?", Seq(value))
+        case sources.GreaterThanOrEqual(attribute, value) => (s"${quoted(attribute)} >= ?", Seq(value))
+        case sources.In(attribute, values)                 =>
```

Comment:
I think we may want to change this (IN Clause) to a parallelized operations later with joinWithCassandraTable or something like that, but we can save that for a later pr.